### PR TITLE
Improve btf_type_match! macro

### PR DESCRIPTION
**Pull request by mendess**
I got better at rust macros and so I came back to improve the `btf_type_match!` macro :smile: 

I think I covered pretty much every use case of a "real" `match` statement, thus making using this macro actually useful instead of annoying.

Please refer to all the tests cases I wrote to test the macro, the goal of the tests is to make sure the code compiles, hence the lack of asserts.

### CODE REVIEW

File: libbpf-rs/src/btf/types.rs [hunk 1]
```diff
@@ -863,26 +844,333 @@ gen_collection_concrete_type! {
...
 /// ```compile_fail
 ///     BtfKind::Int => {
 ///         /* we don't have access to the variable, but we know the scrutinee is an Int */
 ///     }
 /// ```
+///
+/// Multiple Variants
+/// ```compile_fail
+///     BtfKind::Struct | BtfKind::Union => {
+///         /* we don't have access to the variable, but we know the scrutinee is either a Struct or a Union */
```
Comment (danielocfb) [inline libbpf-rs/src/btf/types.rs:857]: Nit: Let's not mix in C style comments and stick with `//` unless there is a good reason not to?

File: libbpf-rs/src/btf/types.rs [hunk 2]
```diff
@@ -863,26 +844,333 @@ gen_collection_concrete_type! {
...
+            _ => false,
+        });
+        btf_type_match!(match ty {
+            BtfKind::Composite(c) => {
+                c.is_struct
+            }
+            _ => false,
+        });
+        // we ask rustfmt to not format this block so that we can keep the trailing `,` in the int
+        // branch.
```
Comment (danielocfb) [inline libbpf-rs/src/btf/types.rs:1107]: Comment seems misleading. There is no `int` branch.

File: libbpf-rs/src/btf/types.rs [hunk 3]
```diff
@@ -863,26 +844,333 @@ gen_collection_concrete_type! {
...
+        dummy_type!(ty);
+
+        match ty.kind() {
+            crate::btf::BtfKind::Int => {
+                let i = crate::btf::types::Int::try_from(ty).unwrap();
+                std::hint::black_box(i);
+                "int"
+            }
+            _ => {
+                std::hint::black_box(());
```
Comment (danielocfb) [inline libbpf-rs/src/btf/types.rs:1128]: What purpose does this `black_box` call have here? It's not clear to me why we need it.

File: libbpf-rs/src/btf/types.rs [hunk 4]
```diff
@@ -863,26 +844,333 @@ gen_collection_concrete_type! {
...
+                $action
+            }
+        }
+             $($rest)*
+        )
+    };
+    // default match case
+    //
+    // we only need the expression case here because this case is not followed by a $rest:tt like
+    // the others, which let's us use the $(,)? pattern.
```
Comment (danielocfb) [inline libbpf-rs/src/btf/types.rs:974]: Well, it could be ;-) Basically we support
```rust
        btf_type_match!(match ty {
            BtfKind::Int => "int",
            _ => "default",
        });
```
but not
```rust
        btf_type_match!(match ty {
            _ => "default",
            BtfKind::Int => "int",
        });
```
which I think is fair and there is no need, just pointing it out.

**Review by danielocfb — APPROVED** [review]
Looks good to me. Thanks for improving this macro! Left a few minor comments.

File: libbpf-rs/src/btf/types.rs (hunk 4 above)
Comment (mendess) [reply libbpf-rs/src/btf/types.rs:974]: yeah that was on purpose, the second code will always generate a warning and in general be incorrect. Technically something the vanilla rust match supports but it seemed hardly worth supporting :grin:

File: libbpf-rs/src/btf/types.rs (hunk 3 above)
Comment (mendess) [reply libbpf-rs/src/btf/types.rs:1128]: I didn't want to use a println but wanted something that looked like a statement. I guess I could have used a simple assignment
```rs
let x = i;
x
```
instead of being all fancy with the black_box.

TL;DR: It serves no special purpose besides being a statement

**Comment by danielocfb** [comment]
Merged. Thanks for the comprehensive set of tests covering the individual cases!
